pkg/test/istio.io/examples/example.go

```diff
@@ -54,16 +54,22 @@ func New(t *testing.T, name string) Example {
 	}
 }
 
-// AddScript adds a directive to run a script
-func (example *Example) AddScript(namespace string, script string, output outputType) {
+// RunScript adds a directive to run a script
+func (example *Example) RunScript(script string, output outputType) {
 	//fullPath := getFullPath(istioPath + script)
 	example.steps = append(example.steps, newStepScript("./"+script, output))
 }
 
-// AddFile adds an existing file
-func (example *Example) AddFile(namespace string, path string) {
+// Apply applies an existing file
+func (example *Example) Apply(namespace string, path string) {
 	fullPath := getFullPath(istioPath + path)
-	example.steps = append(example.steps, newStepFile(namespace, fullPath))
+	example.steps = append(example.steps, newStepFile(namespace, fullPath, false))
+}
+
+// Delete deletes an existing file
+func (example *Example) Delete(namespace string, path string) {
```

Comment:
Sorry, late to the party ... Just trying to understand the use case here.

Will the sequence always be this?

```go
ex.Apply(...)
ex.RunScript(...)
ex.Delete(...)
```

If so, I'm concerned that if the script fails we'll never execute the delete.

Maybe we should change things around a bit so we can specify what we want the run to do up-front:

```go
// First, define an interface for a "thing" to do.
type Task interface {
  Run() error
}

// Next define a composite set of tasks, which is also a Task, itself.
type Tasks []Task

// Verify that Tasks is-a Task
_ Task = Tasks{}

// Execute the tasks in order. Stop when the first error is encountered.
func (tasks Tasks) Run() error {
  for _, task := range tasks {
    if err := task.Run(); err != nil {
      return err
    }
  }
  return nil
}

// Then we can make example control the lifecycle
type Example struct {
  Before Task
  Test Task
  After Task
}

// Execute the tasks in order. Stop when the first error is encountered.
func (e *Example) Run() error {
  // Run the cleanup task at the end, regardless of errors.
  if e.After != nil {
    defer func() { _ = e.After.Run()}()
  }

  if e.Before != nil {
    if err := e.Before.Setup(); err != nil {
      return err
    }
  }

  if err := e.Test.Run(); err != nil {
    return err
  }
}

func (e *Example) RunOrFail(t *testing.T) {
  if err := e.Run(); err != nil {
    t.Fatal(err)
  }
}

// And then in the test code...
func TestAuthzHTTP(t *testing.T) {
  framework.NewTest(t).Run() {
    e := &Example {
      Before: Tasks {
        newBashTask("setup.sh", examples.TextOutput),
        newApplyTask(ns, "samples/bookinfo/platform/kube/bookinfo.yaml"),
        newApplyTask(ns, "samples/bookinfo/networking/bookinfo-gateway.yaml"),
        newApplyTask(ns, "samples/sleep/sleep.yaml"),
        newBashTask("wait.sh", examples.TextOutput),
      },
      After: Tasks {
        newBashTask("clean.sh", examples.TextOutput),
        newDeleteTask(ns, "samples/bookinfo/platform/kube/bookinfo.yaml"),
        newDeleteTask(ns, "samples/bookinfo/networking/bookinfo-gateway.yaml"),
        newDeleteTask(ns, "samples/sleep/sleep.yaml"),
        newDeleteTask("", "samples/bookinfo/platform/kube/rbac/rbac-config-ON.yaml"),
        newDeleteTask(ns, "samples/bookinfo/platform/kube/rbac/productpage-policy.yaml"),
        newDeleteTask(ns, "samples/bookinfo/platform/kube/rbac/details-reviews-policy.yaml"),
        newDeleteTask(ns, "samples/bookinfo/platform/kube/rbac/ratings-policy.yaml"),
      },
      Test: Tasks {
        ...
      },
    } 
    e.RunOrFail(t *testing.T)
  }
}

```

Reply:
There are some cases of 
```golang
ex.Apply()
ex.RunScript()
ex.Clean()
ex.Apply()
ex.RunScript()
ex.Apply()
ex.RunScropt()
ex.Clean()
```
which means it's not a sequential flow and we need to mix the Apply/RunScript/Clean in the flow unless we split it into separate function. 

To always run the clean script, how about let me put it in the main `TestAuthzHTTP()` which seems much cleaner?

Reply:
In that scenario, you have 3 "examples". Each Apply/Run/Delete is one "example".

Reply:
`DocTestCase`?

Reply:
I'm trying to follow the exact same steps as listed on the istio.io. Separating them into different smaller `examples` is okay but I feel it's not preferred:

1) It no longer reflect the original layout of the task in istio.io
2) we need to re-apply some policy multiple times which is different from the flow in the istio.io, for example, the task asks user to: 1) apply policy A, 2) verify effect of A; 3) apply policy B; 4) verify effect of B and A. If splitting to multiple examples, the second example will need to re-apply the policy A again in order to see its effect.

Any thoughts on this?

Reply:
@yangminzhu I agree with you that the point of this exercise is to match the flow of the original doc, so that we'll be able to automatically generate the code in the docs from these tests.

Reply:
@yangminzhu sure that's fair. We can ignore this for now. My concern was that if one test fails, it would not perform the teardown step and thus leave the system in an unstable state causing subsequent tests to fail.  I suppose we can cross that bridge when we get there.